### PR TITLE
Wip/d4s/nova5

### PR DESCRIPTION
**Pull request by d4s**
Refactor Fizz-based support.
Added initial support of Nova 5 headset.
Minor fix in USB Interface ID detection.

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by d4s** [comment]
Since CRC32 command needs clarifying, I commented out that part and got:
```
# fwupdtool --plugins steelseries get-devices
Intel(R) Client Systems NUC8i7HVK
│
├─Aerox 3 Wireless USB Receiver:
│ │   Device ID:          26a3002d823e0907f294dfb781a5794c646c21da
│ │   Current version:    1.11.4
│ │   Vendor:             SteelSeries ApS (USB:0x1038)
│ │   Install Duration:   13 seconds
│ │   GUID:               291b0582-8d08-5714-88db-986911c744d9 ← USB\VID_1038&PID_1838
│ │   Device Flags:       • Updatable
│ │                       • Supported on remote server
│ │                       • Cryptographic hash verification is available
│ │                       • Unsigned Payload
│ │
│ └─Aerox 3 Wireless Mouse via USB Receiver:
│       Device ID:        204083f3da2e1b5c9f5fd31d44454e7117c80323
│       Current version:  1.10.9
│       Vendor:           SteelSeries ApS (USB:0x1038)
│       Install Duration: 38 seconds
│       Battery:          100% (threshold 20%)
│       GUID:             541b2713-d367-5240-a443-bed07f09ffbf ← STEELSERIES\VID_1038&PID_1838&PROTOCOL_FIZZ_TUNNEL
│       Device Flags:     • Updatable
│                         • Supported on remote server
│                         • Cryptographic hash verification is available
│                         • Unsigned Payload
│     
└─Arctis Nova 5 USB Receiver:
  │   Device ID:          84d0e3f2a0f8b2328f7995767b23ebb40494723f
  │   Current version:    0.01.007.000
  │   Vendor:             SteelSeries ApS (USB:0x1038)
  │   Install Duration:   13 seconds
  │   GUID:               ba9a40e9-9992-5673-8cdb-c60a5c1e3ca2 ← USB\VID_1038&PID_2232
  │   Device Flags:       • Updatable
  │                       • Cryptographic hash verification is available
  │                       • Unsigned Payload
  │
  └─Unknown Device:
        Device ID:        1ab7cb96ae3637ab8fee56b8b6609c4af16c4867
        Current version:  000000001.07
        Vendor:           SteelSeries ApS (USB:0x1038)
        Install Duration: 38 seconds
        Battery:          100% (threshold 20%)
        GUID:             a0e5f3d2-6c1c-5478-8ea5-9e118951c07b ← STEELSERIES\VID_1038&PID_2232&PROTOCOL_FIZZ_TUNNEL
        Device Flags:     • Updatable
                          • Cryptographic hash verification is available
                          • Unsigned Payload
```

File: plugins/steelseries/fu-steelseries-fizz-gen1.c [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+				    mode,
+				    error))
+		return NULL;
+
+	fu_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+	if (!fu_steelseries_fizz_gen1_cmd(self, data, sizeof(data), TRUE, error))
+		return NULL;
+	fu_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_TRIPLET);
```
Comment (hughsie) [inline plugins/steelseries/fu-steelseries-fizz-gen1.c:112]: can the version format be set in _init?

**Comment by hughsie** [comment]
> Since CRC32 command needs clarifying

Is it possible I broke it in a1cdbb012447943286e42f4fee5941227dc9dab1 perhaps?

**Comment by hughsie** [comment]
> Unused:  fu_steelseries_fizz_get_version

I think this is CI talking the truth :)

File: plugins/steelseries/fu-steelseries-fizz-gen1.c (hunk 1 above)
Comment (d4s) [reply plugins/steelseries/fu-steelseries-fizz-gen1.c:112]: yes, but clarifying the version format with vendor.
Something strange for raw string format read from the devices:
- dongle: 0.01.006.000, but F/W file is 1.0.6
- headset: 000000001.07, but F/W file is 1.0.7

that's why I set it __temporary__ here. I feel the triplet format should be used everywhere.

**Comment by d4s** [comment]
> > Since CRC32 command needs clarifying
> 
> Is it possible I broke it in [a1cdbb0](https://github.com/fwupd/fwupd/commit/a1cdbb012447943286e42f4fee5941227dc9dab1) perhaps?

I don't thin so, already checked -- the v1 works well, but v2 is needed the bootloader mode for accessing this command.
Removed at all from the setup, since it has no any value here -- should be checked only after flashing.

**Comment by d4s** [comment]
> > Unused:  fu_steelseries_fizz_get_version
> 
> I think this is CI talking the truth :)

yes ;) Still under development.

**Comment by d4s** [comment]
@hughsie need an advice.

The headset might be connected via dongle and via USB-C cable in the same time, please check below.
The sound is routed only via dongle and that expected. 

But the headset itself could be flashed only via USB cable attached directly.

So, I've tried to ask user to attach the cable in commit ad1f3802c841b1efe68d0a7d9d2b4b9dfbcae337 but it is failed with the error:
```
failed to detach: failed to wait for detach replug: device 1ab7cb96ae3637ab8fee56b8b6609c4af16c4867 did not come back
```

I've tried to set the same ID for both devices, based on the unique serial number of the headset, but got the same result. Looks like I need some help here.

The same device is attached via dongle and USB cable directly:
```
Intel(R) Client Systems NUC8i7HVK
│
├─Arctis Nova 5 Headset:
│     Device ID:          1ac52034575b5ba222c476c03af219a7a0e2027a
│     Current version:    1.6.0
│     Vendor:             SteelSeries ApS (USB:0x1038)
│     Install Duration:   1 minute
│     GUID:               3722381d-29fb-572f-895b-8584d94a7a77 ← USB\VID_1038&PID_2230
│     Device Flags:       • Updatable
│                         • Requires a bootloader
│                         • Cryptographic hash verification is available
│                         • Unsigned Payload
│   
└─Arctis Nova 5 USB Receiver:
  │   Device ID:          84d0e3f2a0f8b2328f7995767b23ebb40494723f
  │   Current version:    1.7.0
  │   Vendor:             SteelSeries ApS (USB:0x1038)
  │   Install Duration:   2 minutes
  │   GUID:               ba9a40e9-9992-5673-8cdb-c60a5c1e3ca2 ← USB\VID_1038&PID_2232
  │   Device Flags:       • Updatable
  │                       • Requires a bootloader
  │                       • Cryptographic hash verification is available
  │                       • Unsigned Payload
  │ 
  └─Arctis Nova 5 Headset via USB Receiver:
        Device ID:        1ab7cb96ae3637ab8fee56b8b6609c4af16c4867
        Current version:  1.6.0
        Vendor:           SteelSeries ApS (USB:0x1038)
        Battery:          100% (threshold 20%)
        GUID:             a0e5f3d2-6c1c-5478-8ea5-9e118951c07b ← STEELSERIES\VID_1038&PID_2232&PROTOCOL_FIZZ_TUNNEL
        Device Flags:     • Updatable
                          • Cryptographic hash verification is available
                          • Unsigned Payload
```

**Comment by d4s** [comment]
@hughsie just a gentle reminder. Really need your advice here.

**Comment by hughsie** [comment]
@d4s hey, sorry I missed this. I guess my first thought is that logically the headset should be a child of the receiver, not the other way around. If you want fwupd to wait for both devices, then you need to add WAIT_FOR_REPLUG to both -- I suspect what's happening here is you're setting WAIT_FOR_REPLUG on the headset, and then trying to use "Headset via USB Receiver" when the daemon does the next step -- which could be still busy enumerating.

**Comment by gportay** [comment]
@d4s maybe you could have a look at the bt, I asked the user to plug the mouse using the USB-C cable. I am pretty sure the code is around. But do not forget my memory is pretty bad.

Do you know if that protocol is the one for the keyboard Apex Pro mini?

**Comment by d4s** [comment]
> @d4s hey, sorry I missed this. I guess my first thought is that logically the headset should be a child of the receiver, not the other way around. 

so, even if the receiver is not connected? because the headset supporting BT as well. What would be the next step, but probably I'll start from it.

**Comment by d4s** [comment]
> @d4s maybe you could have a look at the bt, I asked the user to plug the mouse using the USB-C cable. I am pretty sure the code is around. But do not forget my memory is pretty bad.

yes, I've check that and other plugins -- the difference is that Aerox could work only via BT __OR__ USB __OR 2.4, while the headset is detecting (and working) via 2.4 __AND__ USB in the same time. So, if I attach the headset via USB -- it doesn't disappear from the system but stay alive as 2.4 connected device.

> Do you know if that protocol is the one for the keyboard Apex Pro mini?

**Comment by d4s** [comment]
@gportay btw could you try if Aerox 3 mouse detection in BT mode still works for you?
In my current check it is stuck in device version query.

Even with the patch:
```
diff --git a/plugins/steelseries/fu-steelseries-fizz-hid.c b/plugins/steelseries/fu-steelseries-fizz-hid.c
index 0cd4367a7..2c77c0312 100644
--- a/plugins/steelseries/fu-steelseries-fizz-hid.c
+++ b/plugins/steelseries/fu-steelseries-fizz-hid.c
@@ -210,4 +210,6 @@ fu_steelseries_fizz_hid_init(FuSteelseriesFizzHid *self)
        fu_device_set_physical_id(FU_DEVICE(self), "hid");
        fu_device_add_protocol(FU_DEVICE(self), "com.steelseries.fizz");
        fu_device_set_remove_delay(FU_DEVICE(self), 300000); /* 5min */
+       fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_READ);
+       fu_udev_device_add_open_flag(FU_UDEV_DEVICE(self), FU_IO_CHANNEL_OPEN_FLAG_WRITE);
 }
```

**Comment by hughsie** [comment]
> because the headset supporting BT as well. What would be the next step, but probably I'll start from it.

I'm not sure I'd actually add the child mouse to the receiver parent in the case of BT -- it's a laptop->device connection and it'd be a bit weird to include the receiver at all.

**Comment by d4s** [comment]
@hughsie how about review and upstream the current state, since the base support were added months ago?
It allows updating the directly attached dongle and headset, and while we are struggling with several connections, I put a stub to ask user to attach the USB cable and fail the update for a child device:
https://github.com/fwupd/fwupd/pull/7488/files#diff-9d2ff61078ef3b6ed76fc77740a55ffd36cc6a37f0f413e144a0daa846de498dR308

**Review by hughsie — APPROVED** [review]
Looks good -- lets merge this and we can do some rustgen conversion as a 2nd round.

**Pull request merged**